File: extension/src/shared/text-normalizer.ts
import { classifyCodePoint, type Language } from './languages.js';

/**
 * Per-language letter counts for one message. This - never the text - is what leaves the page.
 * Mirrors AutoLang.Core.MessageStats.
 */
export type LetterCounts = Partial<Record<Language, number>>;

export function relevantLetters(counts: LetterCounts): number {
  return Object.values(counts).reduce((sum, n) => sum + (n ?? 0), 0);
}

// Only spans that carry letters while saying nothing about how the writer types need removing.
// Digits, emoji, punctuation and whitespace are already ignored, because a character is counted
// only when it is a letter in a known script.
const SCHEME_URL = /\b(?:https?|ftp):\/\/\S+/gi;
const WWW_URL = /\bwww\.\S+/gi;
const EMAIL = /\S+@\S+\.\S+/g;

// The TLD allowlist is what keeps "ok.thanks" from being read as a domain. A generic
// \w+\.\w+ pattern would eat ordinary text.
const BARE_DOMAIN =
  /\b(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+(?:com|org|net|io|me|co|il|gov|edu|info|biz|app|dev|ai|ly|uk|de|fr)\b(?:\/\S*)?/gi;

export function strip(text: string | null | undefined): string {
  if (!text) return '';
  return text
    .replace(SCHEME_URL, ' ')
    .replace(WWW_URL, ' ')
    .replace(EMAIL, ' ')
    .replace(BARE_DOMAIN, ' ');
}

/**
 * Counts letters per language. Iterating a string with for..of yields whole code points, so
 * surrogate pairs stay intact and emoji are never miscounted as two characters.
 */
export function analyze(text: string | null | undefined): LetterCounts {
  const cleaned = strip(text);
  if (cleaned.length === 0) return {};

  const counts: LetterCounts = {};
  for (const char of cleaned) {
    const language = classifyCodePoint(char.codePointAt(0)!);
    if (language) counts[language] = (counts[language] ?? 0) + 1;
  }
  return counts;
}

File: agent/AutoLang.Core/TextNormalizer.cs
using System.Text.RegularExpressions;

namespace AutoLang.Core;

/// <summary>
/// Turns message text into per-language letter counts.
///
/// PDR section 5 asks to strip URLs, numbers, emoji, punctuation and whitespace. Only the first of
/// those actually needs stripping: because <see cref="ScriptTable.Classify"/> counts a character
/// only when it is a letter in a known script, digits, emoji, punctuation and whitespace are
/// already ignored. URLs and email addresses are different - they are full of Latin letters that
/// would otherwise be read as "this person writes English", which is exactly the false signal the
/// PDR warns about.
/// </summary>
public static partial class TextNormalizer
{
    [GeneratedRegex(@"\b(?:https?|ftp)://\S+", RegexOptions.IgnoreCase)]
    private static partial Regex SchemeUrl();

    [GeneratedRegex(@"\bwww\.\S+", RegexOptions.IgnoreCase)]
    private static partial Regex WwwUrl();

    [GeneratedRegex(@"\S+@\S+\.\S+")]
    private static partial Regex Email();

    // Bare domains such as "google.com" are common in chat. The TLD allowlist keeps this from
    // eating ordinary text like "ok.thanks" - a generic \w+\.\w+ pattern would.
    [GeneratedRegex(@"\b(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+(?:com|org|net|io|me|co|il|gov|edu|info|biz|app|dev|ai|ly|uk|de|fr)\b(?:/\S*)?",
        RegexOptions.IgnoreCase)]
    private static partial Regex BareDomain();

    /// <summary>Removes spans that carry letters but say nothing about the writer's language.</summary>
    public static string Strip(string? text)
    {
        if (string.IsNullOrEmpty(text)) return string.Empty;

        var s = SchemeUrl().Replace(text, " ");
        s = WwwUrl().Replace(s, " ");
        s = Email().Replace(s, " ");
        s = BareDomain().Replace(s, " ");
        return s;
    }

    /// <summary>Counts letters per language, iterating by code point so surrogate pairs stay intact.</summary>
    public static MessageStats Analyze(string? text)
    {
        var cleaned = Strip(text);
        if (cleaned.Length == 0) return MessageStats.Empty;

        var counts = new Dictionary<Language, int>();

        for (int i = 0; i < cleaned.Length; i++)
        {
            int codePoint = char.IsHighSurrogate(cleaned[i]) && i + 1 < cleaned.Length && char.IsLowSurrogate(cleaned[i + 1])
                ? char.ConvertToUtf32(cleaned[i], cleaned[++i])
                : cleaned[i];

            if (ScriptTable.Classify(codePoint) is { } language)
                counts[language] = counts.GetValueOrDefault(language) + 1;
        }

        return new MessageStats(counts);
    }
}
